### PR TITLE
MLv2: port time interval description code

### PR DESCRIPTION
**Pull request by camsaul**
I thought I had already ported this code, and I did, but it turns out there is DUPLICATE code for this stuff in MLv1. It does something ever-so-slightly different. So this is a simple PR to port the second version of this stuff and consolidate it into the stuff I already ported.

For the time being these functions are exported in `metadata.ts`, I think I did the TypeScript wrappers correctly but please double-check. I didn't create a custom type for unit because the FE probably shouldn't be enumerating the available units directly, just getting them from the MLv2 directly. In fact long term I think these functions won't be called directly outside of MLv2 at all; they should be called indirectly with `displayInfo` or something like that. But this is a good interim solution I think until we port more stuff over.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30172)

### CODE REVIEW

File: frontend/src/metabase-lib/queries/utils/query-time.js [hunk 1]
```diff
@@ -139,46 +143,6 @@ export function generateTimeFilterValuesDescriptions(filter) {
   }
 }
 
-function generateTimeIntervalDescription(n, unit) {
```
Comment (camsaul) [inline frontend/src/metabase-lib/queries/utils/query-time.js:142]: So apparently this function (almost) always returned an array, unless `!unit && n === 0`, in which case it returns a plain string. This all seemed very wack. It seemed like most the time it got immediately unpacked, or didn't matter.

**Review by metamben — APPROVED** [review]
> Reviewed 4 of 8 files at r1, 4 of 4 files at r3, 1 of 1 files at r4, 13 of 13 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30172)* status: all files reviewed, 4 unresolved discussions (waiting on @camsaul and @kulyk)

___
*[`src/metabase/lib/temporal_bucket.cljc` line 151 at r4](https://reviewable.io/reviews/metabase/metabase/30172#-NTQ0CvK9wjHnI09bDxZ:-NTQ0CvK9wjHnI09bDx_:undefined) ([raw file](https://github.com/metabase/metabase/blob/fd272533af7bcbffc0b0ef57f1ca9feee5a4f782/src/metabase/lib/temporal_bucket.cljc#L151)):*
> ```Clojure
>        (as-> (str/split (name bucketing) #"-") <>
>          (update (vec <>) 0 str/capitalize)
>          (str/join \space <>))))))
> ```





_Suggestion:_
```Clojure
       (let [[unit & more] (str/split (name bucketing) #"-")]
         (str/join \space (cons (str/capitalize unit) more)))))))
```
___
*[`src/metabase/lib/temporal_bucket.cljc` line 156 at r4](https://reviewable.io/reviews/metabase/metabase/30172#-NTQ2r9kBuLl56Xq9dze:-NTQ2r9kBuLl56Xq9dzf:undefined) ([raw file](https://github.com/metabase/metabase/blob/fd272533af7bcbffc0b0ef57f1ca9feee5a4f782/src/metabase/lib/temporal_bucket.cljc#L156)):*
> ```Clojure
>   (if (number? n)
>     n
>     (condp = (keyword n)
> ```





_Suggestion:_
```Clojure
case
```
___
*[`src/metabase/lib/temporal_bucket.cljc` line 170 at r4](https://reviewable.io/reviews/metabase/metabase/30172#-NTQ3MPz6j0wYdzHfvrd:-NTQ3MPz6j0wYdzHfvre:undefined) ([raw file](https://github.com/metabase/metabase/blob/fd272533af7bcbffc0b0ef57f1ca9feee5a4f782/src/metabase/lib/temporal_bucket.cljc#L170)):*
> ```Clojure
>                   (= unit :day) (i18n/tru "Today")
>                   unit          (i18n/tru "This {0}" (format-bucketing unit))
>                   :else         (i18n/tru "Today"))
> ```





_Suggestion:_
```Clojure
      (zero? n) (if (and unit (not= :day))
                  (i18n/tru "This {0}" (format-bucketing unit))
                  (i18n/tru "Today"))
```

**Comment by metamben** [comment]
Reviewable seems to have messed up some of the suggestions, but I think the intention is clear enough anyway.

File: frontend/src/metabase-lib/metadata.ts [hunk 2]
```diff
@@ -47,3 +47,26 @@ export function groupColumns(columns: ColumnMetadata[]): ColumnGroup {
...
 }
+
+export function describeTemporalUnit(
+  unit: string | null = null,
+  n: number = 1,
+): string {
+  return ML.describe_temporal_unit(n, unit);
+}
+
+type IntervalAmount = number | "current" | "next" | "last";
```
Comment (kulyk) [inline frontend/src/metabase-lib/metadata.ts:58]: FWIW, if there's a chance this type might be used externally or should be shared, it's worth moving it to `metabase-lib/types`. No worries if it's not the case tho

**Review by kulyk — APPROVED** [review]
So wonderful to see tons of FE complexity melt ✨

**Review by camsaul — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30172)* status: 7 of 13 files reviewed, 1 unresolved discussion (waiting on @kulyk and @metamben)

___
*[`frontend/src/metabase-lib/metadata.ts` line 58 at r5](https://reviewable.io/reviews/metabase/metabase/30172#gh-1172520365:-NTtgYF44Stwz6olWw1k:b-xa2caq) ([raw file](https://github.com/metabase/metabase/blob/bbfc2f83ca32391baab3385c67b966aa4be758c3/frontend/src/metabase-lib/metadata.ts#L58)):*
<details><summary><i>Previously, kulyk (Anton Kulyk) wrote…</i></summary><blockquote>

FWIW, if there's a chance this type might be used externally or should be shared, it's worth moving it to `metabase-lib/types`. No worries if it's not the case tho
</blockquote></details>

I think eventually this will go away completely in the near future as the code that builds `relative-datetime` clauses gets moved to MLv2

**Pull request merged**